File: frontend/src/api/loom.ts
import client from './client';

export interface PCAPoint {
  pc1: number;
  pc2: number;
  type_int: number;
  agent: string;
}

export interface LoomEpoch {
  epoch: number;
  type_entropy: number | null;
  D: number | null;
  Q: number | null;
  R: number | null;
  qrc: number | null;
  inquiry_roi: number | null;
  survival_rate: number | null;
  signal_pca: PCAPoint[] | null;
}

export interface LoomData {
  epochs: LoomEpoch[];
  crystallization_epoch: number | null;
  protocol: number;
}

export async function fetchLoomData(runId: number): Promise<LoomData> {
  const { data } = await client.get<LoomData>(`/runs/${runId}/loom`);
  return data;
}

/** Parse a raw run_XX_report.json into LoomData.
 *
 * Handles two formats:
 *  - Legacy flat: { epoch, type_entropy, D, Q, R, qrc, inquiry_roi, ... }
 *  - New nested:  { epoch, inquiry: { type_entropy, type_distribution, ... }, ... }
 *  - Top-level wrapper: { run: {...}, epochs: [...] }
 */
export function parseLoomFile(jsonText: string): LoomData {
  const raw = JSON.parse(jsonText);

  // Unwrap {run, epochs} wrapper if present
  let list: any[];
  if (Array.isArray(raw)) {
    list = raw;
  } else if (raw.epochs) {
    list = raw.epochs;
  } else {
    list = [];
  }

  const epochs: LoomEpoch[] = list.map((m: any) => {
    // New nested format: m.inquiry exists
    if (m.inquiry && typeof m.inquiry === 'object') {
      const inq = m.inquiry;
      const td = inq.type_distribution ?? {};
      return {
        epoch: m.epoch ?? 0,
        type_entropy: inq.type_entropy ?? null,
        D: td.DECLARE ?? null,
        Q: td.QUERY ?? null,
        R: td.RESPOND ?? null,
        qrc: inq.query_response_coupling ?? null,
        inquiry_roi: inq.inquiry_roi ?? null,
        survival_rate: m.survival_rate ?? null,
        signal_pca: m.signal_pca ?? null,
      };
    }
    // Legacy flat format: fields are directly on epoch
    return {
      epoch: m.epoch ?? 0,
      type_entropy: m.type_entropy ?? null,
      D: m.D ?? null,
      Q: m.Q ?? null,
      R: m.R ?? null,
      qrc: m.qrc ?? null,
      inquiry_roi: m.inquiry_roi ?? null,
      survival_rate: m.survival_rate ?? null,
      signal_pca: m.signal_pca ?? null,
    };
  });

  return {
    epochs,
    crystallization_epoch: detectCrystallizationEpoch(epochs),
    protocol: 1,
  };
}

/** Compute crystallization epoch client-side for file-loaded data. */
function detectCrystallizationEpoch(epochs: LoomEpoch[]): number | null {
  let streakStart: number | null = null;
  let count = 0;
  for (const ep of epochs) {
    const te = ep.type_entropy;
    if (te !== null && te < 0.95) {
      if (count === 0) streakStart = ep.epoch;
      count++;
      if (count >= 5 && streakStart !== null) return streakStart;
    } else {
      count = 0;
      streakStart = null;
    }
  }
  return null;
}

File: backend/app/api/loom.py
"""Loom data endpoint — serves epoch metrics filtered for Neural Loom visualisation."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..db import get_db
from ..models import Run, EpochLog

router = APIRouter(prefix="/api/runs", tags=["loom"])


def _find_crystallization_epoch(epochs: list[dict]) -> int | None:
    """First epoch in a streak of 5 consecutive epochs with type_entropy < 0.95."""
    streak_start = None
    streak_count = 0
    for ep in epochs:
        te = ep.get('type_entropy')
        if te is not None and te < 0.95:
            if streak_count == 0:
                streak_start = ep['epoch']
            streak_count += 1
            if streak_count >= 5:
                return streak_start
        else:
            streak_count = 0
            streak_start = None
    return None


@router.get("/{run_id}/loom")
def get_loom_data(run_id: int, db: Session = Depends(get_db)):
    """Return epoch metrics filtered for the Neural Loom scrubber.

    Response shape:
        {
          "epochs": [{
            "epoch": int,
            "type_entropy": float | null,
            "D": float | null, "Q": float | null, "R": float | null,
            "qrc": float | null,
            "inquiry_roi": float | null,
            "survival_rate": float | null,
            "signal_pca": [{pc1, pc2, type_int, agent}] | null
          }],
          "crystallization_epoch": int | null,
          "protocol": int
        }
    """
    run = db.query(Run).get(run_id)
    if not run:
        raise HTTPException(status_code=404, detail="Run not found")

    logs = (
        db.query(EpochLog)
        .filter(EpochLog.run_id == run_id)
        .order_by(EpochLog.epoch)
        .all()
    )

    epochs = []
    for log in logs:
        m = log.metrics
        inq = m.get('inquiry') or {}
        td = inq.get('type_distribution') or {}

        epochs.append({
            'epoch': log.epoch,
            'type_entropy': inq.get('type_entropy'),
            'D': td.get('DECLARE'),
            'Q': td.get('QUERY'),
            'R': td.get('RESPOND'),
            'qrc': inq.get('query_response_coupling'),
            'inquiry_roi': inq.get('inquiry_roi'),
            'survival_rate': m.get('survival_rate'),
            'energy_roi': m.get('energy_roi'),
            'signal_pca': m.get('signal_pca'),
        })

    protocol = run.params.get('protocol', 1)

    return {
        'epochs': epochs,
        'crystallization_epoch': _find_crystallization_epoch(epochs),
        'protocol': protocol,
    }
